### PR TITLE
Get rid of the LEGATE_MAX_CPU_COMMS

### PR DESCRIPTION
**Pull request by eddy16112**
The communicators are created on demand now.

### CODE REVIEW

File: legate/core/communicator.py [hunk 1]
```diff
@@ -17,6 +17,8 @@
 from abc import ABC, abstractmethod, abstractproperty
 from typing import TYPE_CHECKING
 
+import numpy as native_np
```
Comment (magnatelee) [inline legate/core/communicator.py:20]: use `np`

File: legate/core/communicator.py [hunk 2]
```diff
@@ -147,6 +149,12 @@ def _initialize(self, volume: int) -> FutureMap:
             side_effect=True,
         )
         cpucoll_id = task.execute_single()
+        cpucoll_id.wait()
```
Comment (magnatelee) [inline legate/core/communicator.py:152]: this is very bad. why do you need to block on the task? this will stall the whole downstream pipeline.

File: legate/core/communicator.py [hunk 3]
```diff
@@ -147,6 +149,12 @@ def _initialize(self, volume: int) -> FutureMap:
...
+        coll_uid = native_np.frombuffer(
+            cpucoll_id.get_buffer(), dtype=native_np.int32
+        )[0]
+        self._runtime.issue_execution_fence(block=True)
```
Comment (magnatelee) [inline legate/core/communicator.py:156]: this is also very bad. why do you need to block the entire execution here?

File: legate/core/communicator.py [hunk 4]
```diff
@@ -147,6 +149,12 @@ def _initialize(self, volume: int) -> FutureMap:
...
+        self._runtime.core_library.legate_cpucoll_initcomm(coll_uid)
```
Comment (magnatelee) [inline legate/core/communicator.py:157]: hoist this to a task, then you don't need to remove all the previous blocking calls.

File: src/core/comm/coll.cc [hunk 5]
```diff
@@ -50,18 +50,15 @@ enum CollTag : int {
 
 static int mpi_tag_ub = 0;
 
-static std::vector<MPI_Comm> mpi_comms;
+static std::vector<MPI_Comm> mpi_comms(0);
```
Comment (magnatelee) [inline src/core/comm/coll.cc:53]: this feels like an unnecessary micro-optimization that doesn't do any good. I wouldn't mind keeping around the default allocation for this vector even when no cpu communicators are used.

File: src/core/comm/coll.cc [hunk 6]
```diff
@@ -270,16 +264,9 @@ int collInit(int argc, char* argv[])
   CHECK_MPI(MPI_Comm_get_attr(MPI_COMM_WORLD, MPI_TAG_UB, &tag_ub, &flag));
   assert(flag);
   mpi_tag_ub = *tag_ub;
-  // create mpi comms
-  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
-  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i])); }
+  mpi_comms.clear();
```
Comment (magnatelee) [inline src/core/comm/coll.cc:267]: I don't see why this is necessary. is this because you're unsure that the statically declared vector is not empty?

File: src/core/comm/coll.cc [hunk 7]
```diff
@@ -270,16 +264,9 @@ int collInit(int argc, char* argv[])
...
 #else
-  thread_comms.resize(MAX_NB_COMMS);
-  for (int i = 0; i < MAX_NB_COMMS; i++) {
-    thread_comms[i].ready_flag = false;
-    thread_comms[i].buffers    = nullptr;
-    thread_comms[i].displs     = nullptr;
-  }
+  thread_comms.clear();
```
Comment (magnatelee) [inline src/core/comm/coll.cc:269]: same comment here. why is this necessary?

File: src/core/comm/coll.cc [hunk 8]
```diff
@@ -290,34 +277,46 @@ int collFinalize(void)
   assert(coll_inited == true);
   coll_inited = false;
 #ifdef LEGATE_USE_GASNET
-  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
+  for (int i = 0; i < mpi_comms.size(); i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
```
Comment (magnatelee) [inline src/core/comm/coll.cc:280]: let's use a range loop here, instead of a C style loop. (i.e., `for (MPI_Comm& mpi_comm : mpi_comms) { ... }`)

File: src/core/comm/coll.cc (hunk 5 above)
Comment (eddy16112) [reply src/core/comm/coll.cc:53]: I just want to make sure the vector is empty.

File: src/core/comm/coll.cc (hunk 6 above)
Comment (eddy16112) [reply src/core/comm/coll.cc:267]: Yes, I am not sure if the vector will be empty by default. Maybe I can do a check instead of using the clear.

File: legate/core/communicator.py (hunk 1 above)
Comment (eddy16112) [reply legate/core/communicator.py:20]: I just want to distinguish the numpy with cunumeric, so I rename it as native_np.

File: src/core/comm/coll.cc (hunk 6 above)
Comment (eddy16112) [reply src/core/comm/coll.cc:267]: fixed.

File: src/core/comm/coll.cc (hunk 7 above)
Comment (eddy16112) [reply src/core/comm/coll.cc:269]: fixed.

File: src/core/comm/coll.cc (hunk 8 above)
Comment (eddy16112) [reply src/core/comm/coll.cc:280]: fixed, will push it after the sapling is back to normal.

File: legate/core/communicator.py (hunk 4 above)
Comment (eddy16112) [reply legate/core/communicator.py:157]: I removed the `issue_execution_fence`, but kept the `wait`. It is difficult to convert this function into a task because the `MPI_Comm_dup` inside `legate_cpucoll_initcomm` needs to be called by the main thread of each node. Actually the `MPI_Comm_dup` acts like a `MPI_Barrier`, so it will block the calling thread of all nodes, so I do not think it is necessary to convert it into an index task. As long as I remove the `issue_execution_fence`, it will not block the Legion runtime to progress tasks on other threads.

File: legate/core/communicator.py (hunk 1 above)
Comment (magnatelee) [reply legate/core/communicator.py:20]: we don't import cunumeric in legate core, so there's no need for such a distinction.

File: legate/core/communicator.py (hunk 4 above)
Comment (magnatelee) [reply legate/core/communicator.py:157]: > It is difficult to convert this function into a task because the MPI_Comm_dup inside legate_cpucoll_initcomm needs to be called by the main thread of each node.

I don't think this is 100% accurate. I believe what you really want is to call that function only once per node. note that there's no such thing as "main thread" in the legate model. even the python task you're calling that function in isn't really the main thread, but just one of the threads managed by the runtime.

if you want to launch something per node instead of per processor, you just need to adjust the launch domain and update the mapper to do the right thing.

> Actually the MPI_Comm_dup acts like a MPI_Barrier, so it will block the calling thread of all nodes, so I do not think it is necessary to convert it into an index task.

there's a big difference between blocking in one of the leaf tasks and doing it in the main python task; the former only blocks execution of the downstream tasks, which is the right thing to do, but the latter blocks the submission of the downstream tasks and unrolling of the program's control flow, which should be avoided at all cost. I think you should convert it to an index task not to stall the runtime pipeline. it defeats the purpose of deferred execution.

> As long as I remove the issue_execution_fence, it will not block the Legion runtime to progress tasks on other threads.

no this is not true. what `issue_execution_fence(block=True)` really does is that it blocks on a future returned by the fence. so it's basically the same thing as you calling `wait` on a future from the task. so both the blocking on a fence and the `wait` call on a future block the progress of the program.

File: src/core/comm/coll.cc (hunk 5 above)
Comment (magnatelee) [reply src/core/comm/coll.cc:53]: the default constructor empties the vector, so why do you want to do this explicitly?

File: src/core/comm/coll.cc [hunk 9]
```diff
@@ -264,9 +264,9 @@ int collInit(int argc, char* argv[])
   CHECK_MPI(MPI_Comm_get_attr(MPI_COMM_WORLD, MPI_TAG_UB, &tag_ub, &flag));
   assert(flag);
   mpi_tag_ub = *tag_ub;
-  mpi_comms.clear();
+  assert(mpi_comms.empty() == true);
```
Comment (magnatelee) [inline src/core/comm/coll.cc:267]: isn't `empty` already returning a boolean value? what's the point of this comparison?

File: src/core/comm/coll.cc [hunk 10]
```diff
@@ -286,7 +286,7 @@ int collFinalize(void)
     LEGATE_ABORT;
   }
 #else
-  for (int i = 0; i < thread_comms.size(); i++) { assert(thread_comms[i].ready_flag == false); }
+  for (ThreadComm& thread_comm : thread_comms) { assert(thread_comm.ready_flag == false); }
```
Comment (magnatelee) [inline src/core/comm/coll.cc:289]: I'd use a usual bang operator for negation, though this is just a nitpick.

File: src/core/comm/coll.cc [hunk 11]
```diff
@@ -299,10 +299,18 @@ int collGetUniqueId(int* id)
   return CollSuccess;
 }
 
-int collInitComm(int id)
+int collInitComm(void)
```
Comment (magnatelee) [inline src/core/comm/coll.cc:302]: let's not use `void`.

File: src/core/comm/coll.cc [hunk 12]
```diff
@@ -506,5 +515,5 @@ extern "C" {
 
 void legate_cpucoll_finalize(void) { legate::comm::coll::collFinalize(); }
 
-void legate_cpucoll_initcomm(int id) { legate::comm::coll::collInitComm(id); }
+int legate_cpucoll_initcomm(void) { return legate::comm::coll::collInitComm(); }
```
Comment (magnatelee) [inline src/core/comm/coll.cc:518]: again, let's not use `void` in the signature for a function that takes no argument.

**Comment by magnatelee** [comment]
left some nitpicking comments. otherwise looks good.

File: legate/core/communicator.py [hunk 13]
```diff
@@ -140,17 +140,13 @@ def needs_barrier(self) -> bool:
...
 
     def _initialize(self, volume: int) -> FutureMap:
-        task = Task(
-            self._context,
-            self._init_cpucoll_id,
-            tag=self._tag,
-            side_effect=True,
-        )
-        cpucoll_id = task.execute_single()
+        cpucoll_uid = self._runtime.core_library.legate_cpucoll_initcomm()
```
Comment (magnatelee) [inline legate/core/communicator.py:143]: fyi, technically, this code doesn't work when control replication is turned off, but no communicators would work either.
